### DISCUSSION TITLE
Scalable Gatsby Development 404 Page

### DISCUSSION THREAD

**Discussion by wokejacqueline** [gatsbyjs/gatsby#28041] [Help]
## Description

The Gatsby development 404 page includes an _unconstrained_ list of pages, which can cause extremely long querying times. In my project of less than 400 pages, the 404 page takes as long as 15 seconds to load.

I think any of the following would make for an improved developer experience:
1. Constrain the amount of pages listed in the development 404 page, **OR**
2. Remove the list of pages from the development 404 page altogether, **OR**
3. Bypass the development 404 page and use the same page you would for production. (outlined below)

**In your `gatsby-node.js` file**
```
// bypass development 404 page and use production
exports.onCreatePage = ({ page, actions }) => {
  const { deletePage, createPage } = actions

  if (page.path === '/dev-404-page/') {
    deletePage(page)
  } else {
    createPage(page)
  }
}
```

^ Which sort of implies removing the dev 404 page altogether, but I haven't gone that far in my project yet. Is there a reason to have a separate 404 for dev and production?

This issue was actually causing all of my pages to get slower over time when using Algolia's InstantSearch, and I noticed many people using plugins to bypass this page in development for similar reasons, so although it is a small thing, I think it deserves some attention  :) 

---

### Steps to reproduce

Visit a development 404 page (something like `localhost:3000/abcdefg` should do) in any project of +300 pages.

### Expected result

The 404 page should load as fast as any other page.

### Actual result

It loads really slowly in a project with more than a few hundred pages

### Environment

```
System:
    OS: macOS Mojave 10.14.3
    CPU: (8) x64 Intel(R) Core(TM) i7-6920HQ CPU @ 2.90GHz
    Shell: 5.3 - /bin/zsh
  Binaries:
    Node: 10.17.0 - ~/.nvm/versions/node/v10.17.0/bin/node
    npm: 6.11.3 - ~/.nvm/versions/node/v10.17.0/bin/npm
  Languages:
    Python: 2.7.10 - /usr/bin/python
  Browsers:
    Chrome: 86.0.4240.198
    Firefox: 82.0.2
    Safari: 12.0.3
  npmPackages:
    gatsby: ^2.22.20 => 2.22.20
    gatsby-image: ^2.4.6 => 2.4.6
    gatsby-link: ^2.4.5 => 2.4.5
    gatsby-plugin-algolia: ^0.11.1 => 0.11.1
    gatsby-plugin-google-analytics: ^2.3.3 => 2.3.3
    gatsby-plugin-manifest: ^2.4.10 => 2.4.10
    gatsby-plugin-mdx: ^1.2.14 => 1.2.14
    gatsby-plugin-react-helmet: ^3.3.3 => 3.3.3
    gatsby-plugin-robots-txt: ^1.5.1 => 1.5.1
    gatsby-plugin-sass: ^2.3.3 => 2.3.3
    gatsby-plugin-sharp: ^2.6.10 => 2.6.10
    gatsby-plugin-sitemap: ^2.4.4 => 2.4.4
    gatsby-source-filesystem: ^2.3.10 => 2.3.10
    gatsby-transformer-sharp: ^2.5.4 => 2.5.4
  npmGlobalPackages:
    gatsby-cli: 2.12.44
```

**Comment by LekoArts** [comment]
Thank you for opening this!

The PR https://github.com/gatsbyjs/gatsby/pull/13344 already implemented what you propose and this was released in over a year ago. So make sure that you're on the latest version although your info says `gatsby: ^2.22.20 => 2.22.20`. Our own `gatsbyjs.com` site has over 600 sites and with a hard-refresh the site loads in 4s, in a cached state it loads in 2s.

So I don't see any of the problems you describe.

> I noticed many people using plugins to bypass this page in development for similar reasons

Can you link that?
